File: core/src/engine.cpp
#include "irotoiro/engine.hpp"

#include <cassert>
#include <cmath>

namespace irotoiro {
namespace {

std::array<Window, 48> makeWindows() {
  std::array<Window, 48> result{};
  int n = 0;
  auto idx = [](int r, int c) { return static_cast<uint8_t>(r * 5 + c); };

  for (int r = 0; r < 5; ++r) {
    for (int c = 0; c < 3; ++c) {
      result[n++].cells = {idx(r, c), idx(r, c + 1), idx(r, c + 2)};
    }
  }
  for (int c = 0; c < 5; ++c) {
    for (int r = 0; r < 3; ++r) {
      result[n++].cells = {idx(r, c), idx(r + 1, c), idx(r + 2, c)};
    }
  }
  for (int r = 0; r < 3; ++r) {
    for (int c = 0; c < 3; ++c) {
      result[n++].cells = {idx(r, c), idx(r + 1, c + 1), idx(r + 2, c + 2)};
    }
  }
  for (int r = 0; r < 3; ++r) {
    for (int c = 2; c < 5; ++c) {
      result[n++].cells = {idx(r, c), idx(r + 1, c - 1), idx(r + 2, c - 2)};
    }
  }

  assert(n == 48);
  return result;
}

std::array<std::vector<uint8_t>, 25> makeWindowsThrough() {
  std::array<std::vector<uint8_t>, 25> result{};
  const auto& ws = windows();
  for (uint8_t wi = 0; wi < ws.size(); ++wi) {
    for (uint8_t cell : ws[wi].cells) {
      result[cell].push_back(wi);
    }
  }
  return result;
}

bool isAllKind(const GameState& state, const Window& window, CellKind kind) {
  for (uint8_t cell : window.cells) {
    if (state.board[cell].kind != kind) {
      return false;
    }
  }
  return true;
}

}  // namespace

double Rng::next() {
  a_ = a_ + 0x6D2B79F5u;
  uint32_t t = (a_ ^ (a_ >> 15)) * (1u | a_);
  t = (t + ((t ^ (t >> 7)) * (61u | t))) ^ t;
  return static_cast<double>(t ^ (t >> 14)) / 4294967296.0;
}

int mixTile(int a, int b) {
  const int s = a + b;
  if (s == 1) {
    return P;
  }
  if (s == 3) {
    return G;
  }
  return O;
}

const std::array<Window, 48>& windows() {
  static const std::array<Window, 48> kWindows = makeWindows();
  return kWindows;
}

const std::array<std::vector<uint8_t>, 25>& windowsThroughCell() {
  static const std::array<std::vector<uint8_t>, 25> kWindowsThrough = makeWindowsThrough();
  return kWindowsThrough;
}

int handCount(const GameState& state, int player) {
  return state.hand[player][0] + state.hand[player][1] + state.hand[player][2];
}

int bagCount(const GameState& state) {
  return state.bag[0] + state.bag[1] + state.bag[2];
}

int stockCount(const GameState& state, int player) {
  return state.stock[player][0] + state.stock[player][1] + state.stock[player][2];
}

int occupiedCount(const GameState& state) {
  int count = 0;
  for (const Cell& cell : state.board) {
    if (cell.kind != Empty) {
      ++count;
    }
  }
  return count;
}

bool isTerminal(const GameState& state) {
  return stockCount(state, 0) == 0 && stockCount(state, 1) == 0;
}

int drawOneToHand(GameState& state, int player, Rng& rng) {
  if (handCount(state, player) >= 6) {
    return -1;
  }
  const int total = bagCount(state);
  if (total == 0) {
    return -1;
  }

  int r = static_cast<int>(std::floor(rng.next() * static_cast<double>(total)));
  for (int t = 0; t < 3; ++t) {
    if (r < state.bag[t]) {
      --state.bag[t];
      ++state.hand[player][t];
      return t;
    }
    r -= state.bag[t];
  }
  return -1;
}

GameState initialState(Rng& rng) {
  GameState state;
  state.board.fill(Cell{});
  state.stock = {{{4, 4, 4}, {4, 4, 4}}};
  state.hand = {{{0, 0, 0}, {0, 0, 0}}};
  state.bag = {8, 8, 8};
  state.score = {0, 0};
  state.toMove = 0;

  for (int player = 0; player < 2; ++player) {
    for (int n = 0; n < 2; ++n) {
      drawOneToHand(state, player, rng);
    }
  }
  return state;
}

std::vector<Move> legalPlacements(const GameState& state) {
  std::vector<Move> moves;
  const int player = state.toMove;
  for (uint8_t cell = 0; cell < 25; ++cell) {
    if (state.board[cell].kind != Empty) {
      continue;
    }
    for (uint8_t color = 0; color < 3; ++color) {
      if (state.stock[player][color] > 0) {
        moves.push_back(Move{color, cell});
      }
    }
  }
  return moves;
}

Move randomMove(const GameState& state, Rng& rng) {
  const std::vector<Move> moves = legalPlacements(state);
  assert(!moves.empty());
  const int idx = static_cast<int>(std::floor(rng.next() * static_cast<double>(moves.size())));
  return moves[idx];
}

TurnStats applyTurn(GameState& state, Move move, Rng& rng) {
  const int player = state.toMove;
  const int opponent = 1 - player;
  TurnStats stats;

  state.board[move.cell] = Cell{Ohajiki, move.color};
  --state.stock[player][move.color];

  struct AabWindow {
    uint8_t minorityCell = 0;
    uint8_t mix = 0;
  };

  std::vector<uint8_t> drawWindowIndices;
  std::vector<AabWindow> aabWindows;

  for (uint8_t wi : windowsThroughCell()[move.cell]) {
    const Window& window = windows()[wi];
    if (!isAllKind(state, window, Ohajiki)) {
      continue;
    }

    std::array<int, 3> count = {0, 0, 0};
    for (uint8_t cell : window.cells) {
      ++count[state.board[cell].color];
    }

    int distinct = 0;
    for (int n : count) {
      if (n > 0) {
        ++distinct;
      }
    }

    if (distinct == 1 || distinct == 3) {
      drawWindowIndices.push_back(wi);
      continue;
    }

    int minority = -1;
    int majority = -1;
    for (int color = 0; color < 3; ++color) {
      if (count[color] == 1) {
        minority = color;
      } else if (count[color] == 2) {
        majority = color;
      }
    }

    uint8_t minorityCell = 0;
    for (uint8_t cell : window.cells) {
      if (state.board[cell].color == minority) {
        minorityCell = cell;
        break;
      }
    }
    aabWindows.push_back(AabWindow{minorityCell, static_cast<uint8_t>(mixTile(majority, minority))});
  }

  std::array<bool, 25> consumed{};
  consumed.fill(false);
  std::vector<uint8_t> placedTileCells;

  for (const AabWindow& window : aabWindows) {
    if (consumed[window.minorityCell]) {
      continue;
    }
    if (state.hand[player][window.mix] == 0) {
      continue;
    }

    --state.hand[player][window.mix];
    ++state.score[player];
    state.board[window.minorityCell] = Cell{Tile, window.mix};
    consumed[window.minorityCell] = true;
    placedTileCells.push_back(window.minorityCell);
    ++stats.exchanges;
  }

  std::array<bool, 48> seenTileWindows{};
  seenTileWindows.fill(false);
  for (uint8_t placedCell : placedTileCells) {
    for (uint8_t wi : windowsThroughCell()[placedCell]) {
      if (seenTileWindows[wi]) {
        continue;
      }
      const Window& window = windows()[wi];
      if (!isAllKind(state, window, Tile)) {
        continue;
      }

      seenTileWindows[wi] = true;
      if (state.score[opponent] > 0) {
        --state.score[opponent];
        ++state.score[player];
        ++stats.steals;
      } else if (handCount(state, player) < 6 && handCount(state, opponent) > 0) {
        int tile = -1;
        int best = -1;
        for (int k = 0; k < 3; ++k) {
          if (state.hand[opponent][k] > best) {
            best = state.hand[opponent][k];
            tile = k;
          }
        }
        if (tile >= 0 && state.hand[opponent][tile] > 0) {
          --state.hand[opponent][tile];
          ++state.hand[player][tile];
          ++stats.steals;
        }
      }
    }
  }

  for (size_t k = 0; k < drawWindowIndices.size(); ++k) {
    bool gotAny = false;
    for (int n = 0; n < 3; ++n) {
      const int tile = drawOneToHand(state, player, rng);
      if (tile < 0) {
        break;
      }
      gotAny = true;
    }
    if (gotAny) {
      ++stats.drawEvents;
    }
  }

  state.toMove = static_cast<uint8_t>(opponent);
  return stats;
}

std::string boardSignature(const GameState& state) {
  std::string result;
  result.reserve(25);
  for (const Cell& cell : state.board) {
    if (cell.kind == Empty) {
      result.push_back('.');
    } else if (cell.kind == Ohajiki) {
      static constexpr char kOha[] = {'r', 'b', 'y'};
      result.push_back(kOha[cell.color]);
    } else {
      static constexpr char kTile[] = {'P', 'G', 'O'};
      result.push_back(kTile[cell.color]);
    }
  }
  return result;
}

bool checkInvariants(const GameState& state, std::string* error) {
  int boardOhajiki = 0;
  int boardTiles = 0;
  for (const Cell& cell : state.board) {
    if (cell.kind == Ohajiki) {
      ++boardOhajiki;
    } else if (cell.kind == Tile) {
      ++boardTiles;
    }
  }

  int totalStock = 0;
  int totalHand = 0;
  for (int p = 0; p < 2; ++p) {
    for (int c = 0; c < 3; ++c) {
      totalStock += state.stock[p][c];
      totalHand += state.hand[p][c];
    }
  }

  const int totalScore = state.score[0] + state.score[1];
  const int totalBag = bagCount(state);
  if (boardOhajiki + totalStock + totalScore != 24) {
    if (error != nullptr) {
      *error = "ohajiki total invariant failed";
    }
    return false;
  }
  if (boardTiles + totalHand + totalBag != 24) {
    if (error != nullptr) {
      *error = "tile total invariant failed";
    }
    return false;
  }
  for (int p = 0; p < 2; ++p) {
    if (handCount(state, p) > 6) {
      if (error != nullptr) {
        *error = "hand cap invariant failed";
      }
      return false;
    }
  }
  return true;
}

}  // namespace irotoiro

File: web/engine.js
/* ---------- engine (SPEC.md 準拠 / 検証済み参照実装) ---------- */
(function (root, factory) {
  'use strict';
  if (typeof module === 'object' && module.exports) module.exports = factory();
  else root.Irotoiro = factory();
})(typeof globalThis !== 'undefined' ? globalThis : (typeof window !== 'undefined' ? window : this), function () {
  'use strict';
  const OHA={R:0,B:1,Y:2}, TILE={P:0,G:1,O:2};
  function mixTile(a,b){ const s=a+b; if(s===1)return TILE.P; if(s===3)return TILE.G; return TILE.O; }
  const WINDOWS=(function(){ const ws=[],idx=(r,c)=>r*5+c;
    for(let r=0;r<5;r++)for(let c=0;c<3;c++)ws.push([idx(r,c),idx(r,c+1),idx(r,c+2)]);
    for(let c=0;c<5;c++)for(let r=0;r<3;r++)ws.push([idx(r,c),idx(r+1,c),idx(r+2,c)]);
    for(let r=0;r<3;r++)for(let c=0;c<3;c++)ws.push([idx(r,c),idx(r+1,c+1),idx(r+2,c+2)]);
    for(let r=0;r<3;r++)for(let c=2;c<5;c++)ws.push([idx(r,c),idx(r+1,c-1),idx(r+2,c-2)]);
    return ws; })();
  const WINDOWS_THROUGH=(function(){ const m=Array.from({length:25},()=>[]);
    WINDOWS.forEach((w,wi)=>w.forEach(cell=>m[cell].push(wi))); return m; })();
  function cloneState(s){ return {board:s.board.map(c=>c?{kind:c.kind,color:c.color}:null),stock:s.stock.map(a=>a.slice()),hand:s.hand.map(a=>a.slice()),bag:s.bag.slice(),score:s.score.slice(),toMove:s.toMove}; }
  function initialState(rng){ const s={board:new Array(25).fill(null),stock:[[4,4,4],[4,4,4]],hand:[[0,0,0],[0,0,0]],bag:[8,8,8],score:[0,0],toMove:0};
    for(let p=0;p<2;p++)for(let n=0;n<2;n++)drawOneToHand(s,p,rng); return s; }
  const handCount=(s,p)=>s.hand[p][0]+s.hand[p][1]+s.hand[p][2];
  const bagCount=s=>s.bag[0]+s.bag[1]+s.bag[2];
  const stockCount=(s,p)=>s.stock[p][0]+s.stock[p][1]+s.stock[p][2];
  const isTerminal=s=>stockCount(s,0)===0&&stockCount(s,1)===0;
  function drawOneToHand(s,p,rng){ if(handCount(s,p)>=6)return -1; const total=bagCount(s); if(total===0)return -1;
    let r=Math.floor(rng()*total); for(let t=0;t<3;t++){ if(r<s.bag[t]){ s.bag[t]--; s.hand[p][t]++; return t;} r-=s.bag[t]; } return -1; }
  function legalPlacements(s){ const p=s.toMove,mv=[]; for(let cell=0;cell<25;cell++){ if(s.board[cell]!==null)continue;
    for(let c=0;c<3;c++) if(s.stock[p][c]>0) mv.push({color:c,cell}); } return mv; }
  function applyTurn(s,move,rng){ const p=s.toMove,opp=1-p,ev=[];
    s.board[move.cell]={kind:'oha',color:move.color}; s.stock[p][move.color]--;
    ev.push({type:'place',cell:move.cell,player:p,piece:{kind:'oha',color:move.color}});
    const drawWindows=[],aabWindows=[];
    for(const wi of WINDOWS_THROUGH[move.cell]){ const w=WINDOWS[wi]; const a=s.board[w[0]],b=s.board[w[1]],c=s.board[w[2]];
      if(!a||!b||!c||a.kind!=='oha'||b.kind!=='oha'||c.kind!=='oha')continue;
      const cnt=[0,0,0]; cnt[a.color]++;cnt[b.color]++;cnt[c.color]++; const distinct=cnt.filter(x=>x>0).length;
      if(distinct===1||distinct===3){ drawWindows.push(wi); ev.push({type:'line-oha',window:w.slice(),pattern:distinct===1?'same':'diff'}); }
      else { let mn=-1,mj=-1; for(let col=0;col<3;col++){ if(cnt[col]===1)mn=col; else if(cnt[col]===2)mj=col; }
        const mc=w.find(cell=>s.board[cell].color===mn); aabWindows.push({minorityCell:mc,mixTile:mixTile(mj,mn),minorityColor:mn});
        ev.push({type:'line-oha',window:w.slice(),pattern:'pair'}); } }
    const consumed=new Set(),placedTileCells=[];
    for(const ab of aabWindows){ if(consumed.has(ab.minorityCell))continue; if(s.hand[p][ab.mixTile]<=0)continue;
      s.hand[p][ab.mixTile]--; s.score[p]++; s.board[ab.minorityCell]={kind:'tile',color:ab.mixTile};
      consumed.add(ab.minorityCell); placedTileCells.push(ab.minorityCell);
      ev.push({type:'exchange',cell:ab.minorityCell,player:p,scoredColor:ab.minorityColor,tile:ab.mixTile}); }
    const seen=new Set();
    for(const pc of placedTileCells){ for(const wi of WINDOWS_THROUGH[pc]){ if(seen.has(wi))continue; const w=WINDOWS[wi];
      const a=s.board[w[0]],b=s.board[w[1]],c=s.board[w[2]];
      if(!a||!b||!c||a.kind!=='tile'||b.kind!=='tile'||c.kind!=='tile')continue; seen.add(wi);
      if(s.score[opp]>0){ s.score[opp]--; s.score[p]++; ev.push({type:'steal-point',window:w.slice(),from:opp,to:p}); }
      else if(handCount(s,p)<6&&handCount(s,opp)>0){ let t=-1,best=-1; for(let k=0;k<3;k++) if(s.hand[opp][k]>best){best=s.hand[opp][k];t=k;}
        if(t>=0&&s.hand[opp][t]>0){ s.hand[opp][t]--; s.hand[p][t]++; ev.push({type:'steal-tile',window:w.slice(),from:opp,to:p,tile:t}); } } } }
    for(let k=0;k<drawWindows.length;k++){ const got=[]; for(let n=0;n<3;n++){ const t=drawOneToHand(s,p,rng); if(t<0)break; got.push(t);} if(got.length)ev.push({type:'draw',player:p,tiles:got}); }
    s.toMove=opp; return ev; }
  function randomBotMove(s,rng){ const mv=legalPlacements(s); if(!mv.length)return null; return mv[Math.floor(rng()*mv.length)]; }
  function makeRng(seed){ let a=seed>>>0; return function(){ a|=0; a=(a+0x6D2B79F5)|0; let t=Math.imul(a^(a>>>15),1|a); t=(t+Math.imul(t^(t>>>7),61|t))^t; return ((t^(t>>>14))>>>0)/4294967296; }; }
  function boardSignature(s){ let out=''; const oha='rby',tile='PGO'; for(const c of s.board){ out+=c===null?'.':(c.kind==='oha'?oha[c.color]:tile[c.color]); } return out; }
  return {OHA,TILE,mixTile,initialState,legalPlacements,applyTurn,randomBotMove,makeRng,isTerminal,handCount,bagCount,stockCount,cloneState,WINDOWS,WINDOWS_THROUGH,drawOneToHand,boardSignature};
});
